graph/src/blockchain/block_stream.rs

```diff
@@ -403,13 +417,89 @@ pub trait SubstreamsMapper<C: Blockchain>: Send + Sync {
+                                (a, None) => a,
+                                (None, b) => b,
+                                (Some(a), Some(b)) => Some(std::cmp::min(a, b)),
+                            };
+                            format!(" up to {}", b.unwrap_or(0))
+                        } else {
+                            "".to_string()
+                        };
+                        let mlen = stage.modules.len();
+                        let module = if mlen == 0 {
+                            "".to_string()
+                        } else if mlen == 1 {
+                            format!(" ({})", stage.modules[0])
+                        } else {
+                            format!(" ({} +{})", stage.modules[mlen - 1], mlen - 1)
+                        };
+                        if !stages_str.is_empty() {
+                            stages_str.push_str(", ");
+                        }
+                        stages_str.push_str(&format!("#{}{}{}", i, range, module));
+                    }
+                    let stage_str = if len > 0 { 
+                        format!(" Stages: [{}]", stages_str )
+                    } else { 
+                        "".to_string()
+                    };
+                    let mut jobs_str = "".to_string();
+                    let jlen = progress.running_jobs.len();
+                    for i in 0..jlen {
+                        let job = &progress.running_jobs[i];
+                        if !jobs_str.is_empty() {
+                            jobs_str.push_str(", ");
+                        }
+                        let duration_str = format_duration(Duration::from_millis(job.duration_ms));
+                        jobs_str.push_str(&format!("#{} on Stage {} @ {} | +{}|{} elapsed {}", 
+                                                    i, 
+                                                    job.stage, 
+                                                    job.start_block, 
+                                                    job.processed_blocks, 
+                                                    job.stop_block - job.start_block,
+                                                    duration_str));
+                    }
+                    let job_str = if jlen > 0 {
+                        format!(", Jobs: [{}]", jobs_str)
+                    } else {
+                        "".to_string()
+                    };
+                    info!(&logger, "Substreams backend graph_out last block is {}, \
+                                    {} stages up to {}, {} jobs",
+                                    log_data.last_seen_block, len, up_to.unwrap_or(0), jlen);
+                    debug!(&logger, "Substreams backend graph_out last block is {},{}{}", 
+                        log_data.last_seen_block,
+                        stage_str,
+                        job_str,
+                    );
+                    trace!(
```

Comment:
I think we should log conditional to the logger actual level, if possible. What I mean is that if `GRAPH_LOG=trace` then you get the 3 log line where I think it's better to have a single line outputted based on the level.

Can we determine what level is active here and emit only the correct one? Pseudo-code:

```
match logger.level {
   trace: trace!(...),
   debug: debug!(...),
   _: info!(...) 
}
```

This have the added benefits also that in `info`, we would pay a small computation price. 

I would also personally move the formatting code implementation into `impl ModulesProgress { ... }` directly to avoid cluttering `block_stream` with such details.

Reply:
This doesn't seem to be possible with the logging mechanism we currently use (`slog`). There is no reliable way to request current maximum logging level - both `.is_enabled()` and the static fn `slog::__slog_static_max_level() `always return TRACE. I think it's not too bad to have the lower level printed out. Not sure computation would be substantial for debug case compared to the info, worst part are couple more memory allocations.

Reply:
Ok indeed there is nothing reliable on `slog` to achieve that, sad thing. To achieve this, we would need to have a global access to [pub fn logger_with_levels(show_debug: bool, levels: Option<&str>) -> Logger {](https://github.com/graphprotocol/graph-node/blob/866da3f0337d6b33c390a232426781ff76c59229/graph/src/log/mod.rs#L38-L58).

That would make it work as it's where we decide what to set the logger level to and we have no dynamic update. But could be seen as far fetched.

I'm changing my review to approve and let you guys decide what to do about this duplication. Don't forget to look at my comment about moving the formatting code somewhere else then in `block_stream.rs`.

Reply:
I still am not sure how one sets logging to trace level in the `logger_with_levels()` function, not there is a command line switch for that...

Regarding the name/code pollution of the block_stream I completely agree - moved the string generation all to the `SubstreamsLogData` strucure which is now inside `substreams_block_stream.rs`. I feel it belongs there.

Reply:
@mangas here we had a longer discussion on the subject and more context. I feel that printing all 3 levels is not so bad as info is just one line, debug is more or less detailed information, where the trace_id is very different as a structure (a JSON) compared to the debug and it might be useful to have both at this level. WDYT?

Reply:
So looking more into `logger_with_levels()` I'd either have to implement a lite version of [parse_logging_spec](https://github.com/slog-rs/envlogger/blob/b8c7081592b7088b296f46283d896c8f2f4b251c/src/lib.rs#L259) as it is private and combine the result with the `show_debug` flag. Or `std::mem::transmute` the result of `LogBuilder`s `build()` call. Neither seems very elegant to me, so I feel it's not worth it.

Reply:
Closing this as discussed on slack.